source/dub/compilers/compiler.d

```diff
@@ -174,17 +174,19 @@ interface Compiler {
 	/** Compiles platform probe file with the specified compiler and parses its output.
 		Params:
 			compiler_binary =	binary to invoke compiler with
-			args			=	arguments for the probe compilation
+			args			=	arguments for the probe compilation, typically the compiler specific "don't output executables"
+			arch_flags      =   compiler specific flags derived from the user's arch override
 			arch_override	=	special handler for x86_mscoff
+			arch_override_does_something = will be set to true if arch_flags modify the build environment compared to not specifying them.
 	*/
-	protected final BuildPlatform probePlatform(string compiler_binary, string[] args, string arch_override)
+	protected final BuildPlatform probePlatform(string compiler_binary, string[] args, string[] arch_flags, string arch_override, out bool arch_override_does_something)
```

Comment:
I think we should look into simplifying this. Having 3 parameters around arch is a red flag.

Reply:
Let me think about this for a little bit, enough time has passed since I first wrote this and I changed my mind on some implementations details. I'll make the changes that I want and see if I can reduce the number of flags here in order not to have a garbage API.

Reply:
Sorry for the delay - I can also try to take a look. At the end of the day, if this solves a problem you have, I would be happy to go ahead with it to avoid another great delay.

Reply:
The only thing that I care for changing is modifying the algorithm to preserve `-arch x86_64-unknown-linux-gnu` like flags, so fully specifying the target, but try to strip flags like `-arch x86_64` since those just map to `-m64`. I'm just a little worried that stripping the triple will lead to some obscure errors on weird setups but stripping `-m32` and whatnot should always be safe.

I'm not sure I'll get it done by the end of the day and I'm not the one affected by this issue so I also don't care that much for getting this merged ASAP.

Reply:
So I went with another approach. To determine if the platform changes I just call `probePlatform` twice in each compiler's `determinePlatform`.

I've also taken the warning about the architecture not applying from `probePlatform` and put it into `dmd.d` since it seems to be mostly relevant for `x86_mscoff` vs `x86_omf`, which gdc doesn't suffer from. This, along side moving the default probe compiler flags (`-o- -v -c`) into a separate function, reduced the number of parameters to `probePlatform` to only two: the compiler and the arch flags.

Like I've said that I wanted now full triples like `x86_64-unknown-linux-gnu` are not stripped.

Reply:
The new code looks better, my only concern is calling probePlatform twice, do you know how much time that would add to the build time ? I assume the platform file is fairly easily parsed / analyzed but I never actually measured how much it takes on my system.

Reply:
Did a quick check on my system, got min/max/med of 7/11/10

Reply:
> The new code looks better, my only concern is calling probePlatform twice, do you know how much time that would add to the build time ? I assume the platform file is fairly easily parsed / analyzed but I never actually measured how much it takes on my system.

Using a stopwatch to time `determinePlatform` on my system gives, roughly:
| compiler | Without --arch  | With --arch |
| -------- | --------------- | ----------- |
| dmd      | 3.5ms           | 6.5ms       |
| gdc      | 9ms             | 17ms        |
| ldc      | 15ms            | 28ms        |

The old code also invoked the compiler twice to determine if the arch changed, it just did that in the `probePlatform` function in order to reuse the probe file. Timing that probe file generation gives me 0.038ms so, since it keeps the code cleaner I think it's better to simply recall `probePlatform`. I don't think we have a way to check if the architecture changes without reinvoking the compiler, maybe there's a way to do it simply by looking at `compiler -v` and guessing the default architecture from that but it sounds very brittle.